File: scripts/lib/members.mjs
import fs from 'node:fs'
import path from 'node:path'
import yaml from 'js-yaml'
import { memberPath } from '../../src/utils/urn.ts'
import { LEADERSHIP_ROLE_IDS } from '../../src/utils/roles.ts'
import { toISODate, toNullableISODate } from './dates.mjs'
import { loadYamlDir } from './yamlDir.mjs'

export function loadGroups(groupsDir) {
  return loadYamlDir(groupsDir, {
    transform: (data, file) => {
      const id = data.id || file.replace(/\.ya?ml$/, '')
      return {
        ...data,
        id,
        members: [],
        past_members: [],
        convenors: [],
        co_chairs: [],
        managers: [],
      }
    },
  }).reduce((acc, g) => {
    acc[g.id] = g
    return acc
  }, {})
}

export function loadMembers(membersDir, chairMemberId) {
  if (!fs.existsSync(membersDir)) {
    console.warn(`[members] dir not found: ${membersDir}`)
    return { all: {}, chair: null, current: [], past: [], leadership: [] }
  }

  const all = {}
  const current = []
  const past = []
  const leadership = []
  let chair = null

  for (const file of fs.readdirSync(membersDir)) {
    if (!/\.(ya?ml)$/.test(file)) continue
    const fullPath = path.join(membersDir, file)
    const data = yaml.load(fs.readFileSync(fullPath, 'utf8'))
    if (!data) continue
    const id = data['member-id']
    if (!id) continue

    const roleRecords = Array.isArray(data.roles) ? data.roles : []
    const byRoleId = { _all: { in: { _all: [] } } }

    const withGroup = new Set()
    const withoutGroup = new Set()
    for (const r of roleRecords) {
      if (!r || !r.id) continue
      if (r.group) withGroup.add(r.id)
      else withoutGroup.add(r.id)
    }
    const dropIds = new Set([...withGroup].filter(x => withoutGroup.has(x)))

    for (const role of roleRecords) {
      if (!role || !role.id) continue
      const groupId = role.group
      if (!groupId && dropIds.has(role.id)) continue

      byRoleId[role.id] ||= { in: { _all: [] } }
      byRoleId[role.id].in._all.push(role)
      byRoleId._all.in._all.push(role)

      if (groupId) {
        byRoleId[role.id].in[groupId] ||= []
        byRoleId[role.id].in[groupId].push(role)
        byRoleId._all.in[groupId] ||= []
        byRoleId._all.in[groupId].push(role)
      }
    }

    const allRoles = byRoleId._all.in._all
    const isCurrent = data.active === true || allRoles.some(r => r && typeof r === 'object' && r.to == null)
    const isTheChair = chairMemberId === id
    const isInLeadership = allRoles.some(
      r => r && typeof r === 'object' && LEADERSHIP_ROLE_IDS.has(r.id) && r.to == null,
    )

    const member = {
      ...data,
      url: memberPath(id),
      roles: byRoleId,
      is_current: isCurrent,
      is_the_chair: isTheChair,
      is_in_leadership: isInLeadership,
    }
    all[id] = member

    if (isCurrent) {
      if (!isTheChair && !isInLeadership) current.push(id)
      if (isInLeadership) leadership.push(id)
      if (isTheChair) chair = id
    } else {
      past.push(id)
    }
  }

  return { all, chair, current, past, leadership }
}

const CONVENOR_ROLE_IDS = new Set(['convenor', 'co_chair'])

export function attachMembersToGroups(groups, membersAll) {
  for (const group of Object.values(groups)) {
    group.convenor_terms = []
  }
  for (const member of Object.values(membersAll)) {
    const roleRecords = member.roles?._all?.in
    if (!roleRecords) continue
    for (const [groupId, records] of Object.entries(roleRecords)) {
      if (groupId === '_all') continue
      const group = groups[groupId]
      if (!group) continue
      for (const r of records) {
        const isCurrent = r.to == null
        if (isCurrent) {
          if (r.id === 'manager') group.managers.push(member['member-id'])
          else if (r.id === 'convenor') group.convenors.push(member['member-id'])
          else if (r.id === 'co_chair') group.co_chairs.push(member['member-id'])
          else if (r.id === 'member') group.members.push(member['member-id'])
        } else if (r.id === 'member') {
          group.past_members.push(member['member-id'])
        }
        if (CONVENOR_ROLE_IDS.has(r.id)) {
          const from = toISODate(r.from)
          if (!from) continue
          const to = toNullableISODate(r.to)
          const seat = Array.isArray(r.seat) ? r.seat : (r.seat ? [r.seat] : undefined)
          group.convenor_terms.push({
            member_id: member['member-id'],
            name: member.name,
            from,
            to,
            current: isCurrent,
            role: r.id,
            seat,
          })
        }
      }
    }
  }
  for (const group of Object.values(groups)) {
    group.convenor_terms.sort((a, b) => a.from.localeCompare(b.from) || a.name.localeCompare(b.name))
  }
}

export function enrichConvenorTerms(groups) {
  for (const group of Object.values(groups)) {
    if (!group.convenor_terms) continue
    const events = group.history?.events
    if (!events) continue
    const appointmentByMember = new Map()
    for (const ev of events) {
      if (ev.type !== 'convenor_appointed' && ev.type !== 'convenor_extended') continue
      if (!ev.person_member_id) continue
      const prev = appointmentByMember.get(ev.person_member_id)
      if (!prev || ev.date > prev.date) {
        appointmentByMember.set(ev.person_member_id, ev)
      }
    }
    for (const term of group.convenor_terms) {
      const ev = appointmentByMember.get(term.member_id)
      if (!ev) continue
      if (!term.resolution_ref && ev.resolution_ref) term.resolution_ref = ev.resolution_ref
      if (ev.term_until && (!term.to || ev.term_until > term.to)) {
        term.to = ev.term_until
        const today = new Date().toISOString().slice(0, 10)
        if (ev.term_until < today) term.current = false
      }
    }
  }
}

export function buildOrganizationMembers(nationalBodies, liaisons, membersAll) {
  const shortNameMap = {}
  for (const org of [...nationalBodies, ...liaisons]) {
    if (org.short_name) shortNameMap[org.short_name.toLowerCase()] = org.id
  }
  const result = {}
  for (const [memberId, member] of Object.entries(membersAll)) {
    const aff = member.affiliation
    if (!aff) continue
    const shortName = aff.split(',')[0].trim().replace(/-/g, ' ')
    const orgId = shortNameMap[shortName.toLowerCase()]
    if (orgId) {
      (result[orgId] ||= []).push(memberId)
    }
  }
  return result
}

File: attic/_jekyll-legacy/_plugins/members.rb
module Jekyll

  class Site

    def read_groups
      # Read groups from _data/groups/ YAML (migrated from .adoc collection)
      raw_groups = self.data['groups'] || {}

      self.data['groups'] = {}

      raw_groups.each do |id, group_data|
        self.data['groups'][id] = group_data.merge({
          'members' => [],
          'past_members' => [],
          'convenors' => [],
          'co_chairs' => [],
          'managers' => [],
        })
      end
    end

    def read_members
      # Read members from _data/members/ YAML files
      raw_members = self.data['members'] || {}
      committee = self.config['committee']

      self.data['members'] = {
        'all' => {},
        'chair' => {},
        'current' => [],
        'past' => [],
        'leadership' => [],
      }

      leadership_role_ids = ['chair', 'technical_programme_manager', 'committee_manager', 'convenor', 'co_chair']

      raw_members.each do |id, member_data|
        role_records = member_data['roles']

        is_the_chair = committee['chair'] == id

        # Build role index
        by_role_id = {
          '_all' => {
            'in' => {
              '_all' => [],
            },
          },
        }

        # Deduplicate: if a role_id has entries both with and without groups,
        # drop the groupless one (it's redundant)
        role_ids_with_group = Set.new
        role_ids_without_group = Set.new

        if role_records and role_records.respond_to?('each')
          role_records.each do |role_record|
            role_id = role_record['id']
            next unless role_id
            if role_record['group']
              role_ids_with_group.add(role_id)
            else
              role_ids_without_group.add(role_id)
            end
          end

          drop_ids = role_ids_with_group & role_ids_without_group

          role_records.each do |role_record|
            role_id = role_record['id']
            if role_id != nil
              group_id = role_record['group']

              # Skip groupless role if this role_id also has group entries
              if group_id.nil? && drop_ids.include?(role_id)
                next
              end

              role_timespan = {
                'from' => role_record['from'],
                'to' => role_record['to'],
              }

              by_role_id[role_id] ||= { 'in' => { '_all' => [] } }
              by_role_id[role_id]['in']['_all'] << role_record
              by_role_id['_all']['in']['_all'] << role_record

              if group_id
                if role_timespan['to'] == nil
                  if role_id == 'manager'
                    self.data['groups'][group_id]['managers'] << id
                  elsif role_id == 'convenor'
                    self.data['groups'][group_id]['convenors'] << id
                  elsif role_id == 'co_chair'
                    self.data['groups'][group_id]['co_chairs'] << id
                  elsif role_id == 'member'
                    self.data['groups'][group_id]['members'] << id
                  end
                elsif role_id == 'member'
                  self.data['groups'][group_id]['past_members'] << id
                end

                by_role_id[role_id]['in'][group_id] ||= []
                by_role_id[role_id]['in'][group_id] << role_record
                by_role_id['_all']['in'][group_id] ||= []
                by_role_id['_all']['in'][group_id] << role_record
              end
            end
          end
        else
          role_records = [member_data['role']]
          by_role_id['_all']['in']['_all'] = role_records
        end

        all_roles = by_role_id['_all']['in']['_all']
        is_current = member_data['active'] == true or all_roles.select { |role|
          if not role.is_a?(String)
            role['to'] == nil
          end
        }.size > 0

        is_in_leadership = all_roles.any? { |role|
          if role.is_a?(Hash)
            leadership_role_ids.include?(role['id']) && role['to'] == nil
          end
        }

        parsed_member = member_data.clone
        parsed_member['url'] = "/members/#{id}/"
        parsed_member['roles'] = by_role_id
        parsed_member['is_current'] = is_current
        parsed_member['is_the_chair'] = is_the_chair
        parsed_member['is_in_leadership'] = is_in_leadership

        self.data['members']['all'][id] = parsed_member

        if is_current
          if not is_the_chair and not is_in_leadership
            self.data['members']['current'] << id
          end

          if is_in_leadership
            self.data['members']['leadership'] << id
          end

          if is_the_chair
            self.data['members']['chair'] = id
          end

        else
          self.data['members']['past'] << id
        end
      end
    end

    def read_organization_members
      nb_data = self.data['national_bodies'] || []
      lia_data = self.data['liaisons'] || []

      # Build case-insensitive short_name -> id map from all orgs
      short_name_map = {}
      nb_data.each do |org|
        next unless org['short_name']
        short_name_map[org['short_name'].downcase] = org['id']
      end
      lia_data.each do |org|
        next unless org['short_name']
        short_name_map[org['short_name'].downcase] = org['id']
      end

      # Group members by their organization's id
      org_members = {}
      self.data['members']['all'].each do |member_id, member_data|
        aff = member_data['affiliation']
        next unless aff
        short_name = aff.split(',').first.strip.gsub('-', ' ')
        org_id = short_name_map[short_name.downcase]
        if org_id
          org_members[org_id] ||= []
          org_members[org_id] << member_id
        end
      end

      self.data['organization_members'] = org_members
    end

    def read_projects
      # Load directly from _data/projects/*.yaml
      projects_dir = File.join(self.source, '_data', 'projects')
      yaml_projects = {}
      Dir.glob(File.join(projects_dir, '*.{yaml,yml}')).each do |f|
        data = YAML.safe_load(File.read(f), permitted_classes: [Date, DateTime])
        if data.is_a?(Hash) && data['id']
          yaml_projects[data['id']] = data
        end
      end

      if yaml_projects.any?
        self.data['projects'] = yaml_projects
      else
        # Fallback to collection (for development)
        if self.collections['projects']
          projects = self.collections['projects'].docs.map(&:clone)
          self.data['projects'] = {}
          projects.each do |project|
            self.data['projects'][project.data['id']] = project
          end
        else
          self.data['projects'] = {}
        end
      end
    end

    def generate_member_pages
      self.data['members']['all'].each do |id, member_data|
        page = PageWithoutAFile.new(self, self.source, '', "members/#{id}/index.html")
        page.data['layout'] = 'member'
        page.data['title'] = member_data['name']
        page.data['member'] = member_data
        page.data['member-id'] = id
        # Expose processed member fields directly on page
        member_data.each do |key, val|
          page.data[key] = val
        end
        self.pages << page
      end
    end
  end

  class MemberDataReader < Generator
    safe true
    priority :high

    def generate(site)
      site.read_groups
      site.read_members
      site.read_organization_members
      site.read_projects
      site.generate_member_pages
    end
  end

end
